state/proxy/shadow.go

```diff
@@ -703,16 +714,17 @@ func (s *shadowVmStateDb) Selfdestruct6780(addr common.Address) {
 }
 
 func (s *shadowVmStateDb) GetStorageRoot(addr common.Address) common.Hash {
-	return s.getHash("GetStorageRoot", func(s state.VmStateDB) common.Hash { return s.GetStorageRoot(addr) }, addr)
-}
+	// todo call both return non-empty
+	primeRoot := s.prime.GetStorageRoot(addr)
+	shadowRoot := s.shadow.GetStorageRoot(addr)
 
-func (s *shadowVmStateDb) SetTransientState(addr common.Address, key common.Hash, value common.Hash) {
-	s.run("SetTransientState", func(s state.VmStateDB) error {
-		s.SetTransientState(addr, key, value)
-		return nil
-	})
-}
+	if primeRoot == shadowRoot {
+		return primeRoot
+	}
```

Comment:
If the results of GetStorageHash diverge, an error should be thrown, no? 

It not quite clear to me what this condition and the one from line 725 suppose to do.  From the coding logic, it reads to me that as long as `shadowRoot` is not an empty hash, then always returns  `shadowRoot`; otherwise returns `primeRoot`. 

The condition `if primeRoot == shadowRoot` is not needed and can be handled by line 725 to 729. If `primeRoot` and `shadowRoot` holds the same value, either `shadowRoot` or `primeRoot` can be returned. Line 726 returns the case where the root is not an empty hash and line 729 returns the case where toot is an empty hash.

Reply:
The thing is that we dont want to compare this. Rn `Carmen` does not have this implemented. It seems like this wont ever get implemented because there shouldnt be scenario on our chaIN where this func is called. Since carmens implementation of `StateDb` always returns an empty hash and geth implementation just calls the internal geth func, it can diverge even though its not an error.

`ShadowDb` `GetStorageRoot` implementation is needed mainly for `eth-tests` where it could potentionally be called although as I written above, real scenario on chain wont ever happen.

Reply:
Got it. Would you remove line 721 - 723? This condition is already handled by the next if statement in line 725.